### PR TITLE
added the hinge_embedding_loss

### LINKED ISSUE #21076: hinge_embedding_loss
- [hinge_embedding_loss] #21077

### PR DESCRIPTION
**Pull request by sushmanthreddy**
close #21076

### CODE REVIEW

**Comment by sushmanthreddy** [comment]
@MahmoudAshraf97 can u pls review th pulll request pls??

**Comment by sushmanthreddy** [comment]
@MahmoudAshraf97 @YushaArif99 can u pls review my pull request, i have added the support for hinge_embeddding_loss ,and tests are passing in local system and the sc is attached over here..pls review and let me knew any changes are needed.

**Comment by sushmanthreddy** [comment]
checked supporting datatypes by this method
Screenshot from 2023-08-20 02-30-25

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
         min_dim_size=3,
     ),
     dtype_and_log_input=helpers.dtype_and_values(
-        available_dtypes=helpers.get_dtypes("float"),
+        available_dtypes=helpers.get_dtypes("int"),
```
Comment (MahmoudAshraf97) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:22]: revert this change

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py [hunk 2]
```diff
@@ -57,3 +57,52 @@ def test_log_poisson_loss(
...
+@handle_test(
+    fn_tree="functional.ivy.experimental.hinge_embedding_loss",
+    dtype_and_input=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("float"),
+        min_value=-3,
+        max_value=3,
+        allow_inf=False,
+        min_num_dims=1,
+        max_num_dims=3,
+        min_dim_size=3,
```
Comment (MahmoudAshraf97) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:72]: remove these unnecessary constraints, the test should be as unconstrained as possible
```suggestion

```

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py [hunk 3]
```diff
@@ -57,3 +57,52 @@ def test_log_poisson_loss(
...
+    ),
+    dtype_and_target=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("float"),
+        min_value=-1,
+        max_value=1,
+        allow_inf=False,
+        min_num_dims=1,
+        max_num_dims=3,
+        min_dim_size=3,
```
Comment (MahmoudAshraf97) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:81]:
```suggestion

```

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py [hunk 4]
```diff
@@ -57,3 +57,52 @@ def test_log_poisson_loss(
...
+    ),
```
Comment (MahmoudAshraf97) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:82]: you can also generate two arrays with the same strategy instead of having two strategies

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py [hunk 5]
```diff
@@ -57,3 +57,52 @@ def test_log_poisson_loss(
...
+    margin=st.floats(min_value=0, max_value=3, allow_infinity=False),
+    test_with_out=st.just(False),
```
Comment (MahmoudAshraf97) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:84]: why is this hardcoded to false?
```suggestion
```

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py
Comment (MahmoudAshraf97) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:1]: please address the concerns in the testing review, also add testing for reduction argument

**Review by MahmoudAshraf97 — CHANGES_REQUESTED** [review]
Hello @sushmanthreddy , the function argument should be unified across all backends and still satisfy the superset behaviour, please address the issues mentioned in the review

**Comment by sushmanthreddy** [comment]
@MahmoudAshraf97 I have implemented the changes u have mentioned.And ,I have  added the reduction parameter too in test file.It just got out of sight.
And after reading superset behaviour ,have understood the backend implementation too ,and removed unecessary parameters which are not present in other backends.
One strategy have true labels and other have targets which are binary  1 and -1 ,so it better to have two different strategies ,now tests are passing.can u pls review it ,if changes are needed pls let me knew otherwise pls merge it

**Comment by sushmanthreddy** [comment]
Screenshot from 2023-08-20 20-28-08
tests passing screenshot

File: ivy/functional/backends/paddle/experimental/losses.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    margin: Optional[float] = 1.0,
+    reduction: Optional[str] = "mean",
+    out: Optional[paddle.Tensor] = None,
+) -> paddle.Tensor:
+    return F.hinge_embedding_loss(
+        input,
+        target,
+        margin=margin,
+        reduction=reduction,
+        out=out,
```
Comment (MahmoudAshraf97) [inline ivy/functional/backends/paddle/experimental/losses.py:42]: paddle function doesn't have an out argument

File: ivy/functional/backends/torch/experimental/losses.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    margin: Optional[float] = 1.0,
+    reduction: Optional[str] = "mean",
+    out: Optional[torch.Tensor] = None,
+) -> torch.Tensor:
+    return torch.nn.functional.hinge_embedding_loss(
+        input,
+        target,
+        margin=margin,
+        reduction=reduction,
+        out=out,
```
Comment (MahmoudAshraf97) [inline ivy/functional/backends/torch/experimental/losses.py:37]: torch function doesn't have an out argument

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py [hunk 8]
```diff
@@ -57,3 +57,49 @@ def test_log_poisson_loss(
...
+        test_flags=test_flags,
+        backend_to_test=backend_fw,
+        fn_name=fn_name,
+        on_device=on_device,
+        input=input_data[0],
+        target=target_data[0],
+        margin=margin,
+        reduction=reduction,
+        atol_=1e-2,
+        rtol_=1e-2,
```
Comment (MahmoudAshraf97) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:104]: this is a relatively high tolerance please remove it

**Review by MahmoudAshraf97 — CHANGES_REQUESTED** [review]

**Review by MahmoudAshraf97 — CHANGES_REQUESTED** [review]
Hi, please remove all tolerances not just relative
the out argument doesn't exist in the frameworks but it should stay in the function signature, please check how `handle_out_argument` works and make sure it wraps the main function

**Comment by YushaArif99** [comment]
Hi, @sushmanthreddy. I'll be closing this PR has it has been inactive for over a month now. Thanks! 🙂
